rfcs/45-stageless.md

```diff
@@ -0,0 +1,1078 @@
+   2. In the vast majority of cases, this is the desired behavior. Unless you are using interior mutability (or accessing something outside of the ECS), systems that are compatible will always be **commutative**: their ordering doesn't matter, and these constraints behave "as-if" they were strict.
+   3. Note that if-needed ordering constraints are transitive. `a.before(b)` and `b.before(c)` implies `a.before(c)`. This behavior, while intuitive, can have some unexpected effects: it occurs even if the chain is broken by a **spurious** constraint (in which the two systems are compatible).
+
+Applying an ordering constraint to or from a label causes a ordering constraint to be created between all individual members of that label.
+If an ordering is defined relative to a non-existent system or an unused label, it will have no effect, emitting a warning.
+This relatively gentle failure mode is important to ensure that plugins can order their systems with relatively strong assumptions that the default system labels exist, but continue to (mostly) work if those systems or labels are not present.
+
+In addition to the `.before` and `.after` methods, you can use **system chains** to create very simple linear dependencies between the successive members of an array of systems.
+(Note to readers: this is not the same as "system chaining" in Bevy 0.6 and earlier: that concept has been renamed to "system handling".)
+
+```rust
+fn main(){
+   App::new()
+   // This systems are connected using a string of if-needed ordering constraints
+   .add_system_chain((compute_attack, 
+                      compute_defense,
+                      check_for_crits,
+                      compute_damage,
+                      deal_damage,
+                      check_for_death)
+                      // We can configure and label all systems in the chain at once
+                      .label(GameLabel::Combat))
+   .run()
+}
+```
+
+#### Ordering with `Commands`
+
+Commands (commonly used to spawn and despawn entities or add and remove components) do not take effect immediately.
+Instead, you a copy of the `flush_commands` system must run.
+This **exclusive system** (meaning, it can modify the entire `World` in arbitrary ways and cannot be run in parallel with other systems) collects all created commands and applies them to the `World`.
+
+This pattern is so common that a special form of ordering constraint exists for it: **command-flushed ordering constraints**.
+If system `A` is `before_and_flush` system `B`, the schedule will be unsatisfiable unless there is an intervening `flush_commands` system.
+Note that **this does not insert new copies of a `flush_commands` system**: instead, it functions like an `assert!` statement.
+It has no direct effect, and is merely used to verify that your schedule has been set up correctly according the specified constraint.

```

Comment:
Just a small comment: I'm a bit confused on the purpose of `before_and_flush` (and others). If it doesn't add anything, is it meant to just help you catch bugs earlier on by panicking? Or is it for plugin authors to enforce a command flush before running their systems?

I think the terminology implies that it runs before the system _and_ performs the flush. So maybe some clarity on when to use it and why?

Reply:
> . If it doesn't add anything, is it meant to just help you catch bugs earlier on by panicking? Or is it for plugin authors to enforce a command flush before running their systems?

Both. I agree that it needs more clarity here.

Reply:
Clarified!